src/lib.rs

```diff
@@ -118,8 +116,10 @@ impl Default for Config {
 pub fn init(config: Config) -> Peripherals {
```

Comment:
Hummm. doesn't that just make this Peripherals::init unsafe now? 

I think maybe a "correct" solution is to have 
- `static mut CONFIGURED: AtomicBool`
-  Swap it to true and assert the old value being false
-  When Peripherals are dropped, we consider it "un-configures" and set the bool back.

Reply:
Need to think a bit more about safety of this. but I guess technically this is not making thing any worse 🤷

Reply:
embassy uses a pretty similar solution: init calls Peripherals::take() https://github.com/embassy-rs/embassy/blob/b0c7fa07b27be9025c5e986e339899cdfc9a01e2/embassy-stm32/src/lib.rs#L423 which comes from here https://github.com/embassy-rs/embassy/blob/b0c7fa07b27be9025c5e986e339899cdfc9a01e2/embassy-hal-internal/src/macros.rs#L65 which we should already have. Not sure I'd bother worrying about what happens when peripherals is dropped, since you only really get it at the beginning of the program and if you need it again it usually means you should have borrowed from it rather than taking. there's always steal() anyway

Reply:
Technically it is allowed for you to reconfigure the core with different clocks etc aka re-init

Reply:
if that is supposed to be possible then I agree setting and unsetting on drop should do it. curious what your use-case is for reconfiguring? also not sure if that static bool from the macro will be hanging around even if it's unused. (well, don't lose any sleep over 1 byte of ram 😆 )

actually, I think at least the RCC init in embassy-stm32 are already supposed to be able to cope with already-configured clocks in case of e.g. a bootloader. if all the inits are "only call when nothing is being used" rather than "only call once, ever", no reason you wouldn't be able to init() again in the same app. possibly problematic with time-driver?
might even be worth upstreaming a drop impl for struct Peripheral that unsets that flag

Reply:
...wait, i've just realized that we're already doing Peripherals::take(). only difference vs embassy-stm32 is that we do it after calling all the unsafe init()s, so if we wanted to use that to stop us from doing anything unsafe, the change should just be to call that first and then return it at the end like embassy-stm32 does

Reply:
Oh I don't have a real use case right now, but hypothetically (very hypothetical) that for example low power application might opt to use the internal oscillator when system is in low power state and run at a lower clock. I think I'm just being pedantic here.

Reply:
But you are right, good catch. I think in this case all we need to do is just move the taking peripheral step earlier it would be all good.